public/js/gulliver.js

```diff
@@ -0,0 +1,95 @@
+/* eslint-env browser */
+
+/**
+ * Note: window.gapiReady must be a promise that resolves to
+ * window.gapi.
+ *
+ * @template T
+ * @param {string} name the library to load
+ * @return {Promise<T>}
+ */
+function gapiLoad(name) {
+  return new Promise((resolve, reject) => {
+    if ('gapiReady' in window) {
+      window.gapiReady.then(gapi => {
+        gapi.load(name, () => resolve(gapi[name]));
+      });
+    } else {
+      reject('Promise window.gapiReady not found');
+    }
+  });
+}
+
+/**
+ * Promise'd version of [`gapi.client.load`](https://developers.google.com/api-client-library/javascript/reference/referencedocs#gapiclientloadname--------version--------callback).
+ *
+ * @template T
+ * @param {string} name the API client to load
+ * @param {string} [version="v1"] version
+ * @return {Promise<T>} Promise resolving to gapi.client[name]
+ */
+function clientLoad(name, version) {
+  version = version ? version : 'v1';
+  return gapiLoad('client').then(client => {
+    return new Promise(resolve => {
+      client.load(name, version, () => resolve(client[name]));
+    });
+  });
+}
+
+/**
+ * Promise'd version of [`gapi.auth2.init`](https://developers.google.com/identity/sign-in/web/reference#gapiauth2initparams).
+ *
+ * @param {any} params https://developers.google.com/identity/sign-in/web/reference#gapiauth2initparams
+ * @return {Promise<gapi.auth2.GoogleAuth>} Promise resolving to an initialized gapi.auth2.GoogleAuth object
+ */
+function authInit(params) {
+  return gapiLoad('auth2').then(auth2 => {
+    /* Ideally we'd just return `auth2.init(params)` here, but
+     * instead we need to work around a few bugs and surprises in
+     * `auth2.init()` and the "Promise" it returns.
+     */
+    return new Promise(resolve => {
+      auth2.init(params).then(t => {
+        t.then = null;
+        resolve(t);
+      });
+    });
+  });
+}
+
+(() => {
+  const [login, logout] = [document.getElementById('login'), document.getElementById('logout')];
+
+  /**
+   * @param {GoogleUser} user
+   */
+  function onChange(user) {
+    if (user.isSignedIn()) {
+      // @TODO Send id_token to backend
+      console.log('id_token', user.getAuthResponse().id_token);
+      login.disabled = true;
+      logout.disabled = false;
+    } else {
+      console.log('user signed out/never signed in');
+      login.disabled = false;
+      logout.disabled = true;
+    }
+  }
+
+  /* eslint-disable camelcase */
+  const params = {
+    scope: 'email',
+    client_id: '605287872172-js6omne47i1k79hnfo7d4bdu9rlemslr.apps.googleusercontent.com',
```

Comment:
I don't think there's a standard way, just a bunch of options.

I guess the client_id should ultimately come from
https://github.com/GoogleChrome/gulliver/blob/master/config/config.json
though I don't know of a nice way to propagate it to the frontend.

Some possibilities:

1. Template the JS file itself, and run through templating system when
publishing. (Unexpected, and templating code is nasty.)
1. Load the config.json from the JS (synchonously). (Slow, maybe okay for
now.)
1. Make a templated/generated config.js (basically just returns
config.json), load via blocking `<script>`. (Maybe okay for now.)
1. Add a small chunk of config JS to the top of every HTML page. (Messy,
need to make inline JS executable.)
1. Add a module system, then do something similar to the NodeJS side
(`const config = require('config')`). (Needs module system. Probably the
"right" way, but needs a module system.)


On Mon, Jul 18, 2016 at 12:09 PM, André Cipriani Bandarra <
notifications@github.com> wrote:

> In public/js/gulliver.js
> <https://github.com/GoogleChrome/gulliver/pull/19#discussion_r71117491>:
>
> > +    if (user.isSignedIn()) {
> > +      // @TODO Send id_token to backend
> > +      console.log('id_token', user.getAuthResponse().id_token);
> > +      login.disabled = true;
> > +      logout.disabled = false;
> > +    } else {
> > +      console.log('user signed out/never signed in');
> > +      login.disabled = false;
> > +      logout.disabled = true;
> > +    }
> > +  }
> > +
> > +  /* eslint-disable camelcase */
> > +  const params = {
> > +    scope: 'email',
> > +    client_id: '605287872172-js6omne47i1k79hnfo7d4bdu9rlemslr.apps.googleusercontent.com',
>
> The client_id and secret look like something that should be stored in a
> config file. A sample of the config file can be saved in the repo and users
> will have to change and add their own key/secret to run the project. What
> are the best practices to this in Node? I'm guessing either a JSON file
> or a config.js that creates globals?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleChrome/gulliver/pull/19/files/5ca8d86f3ba0fb2c5b0a27abf03ef757bbaf0825#r71117491>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADILPyyTHgA1Fo-7FLuFHD_aXda-Q_Eks5qW0KzgaJpZM4JNUDy>
> .
>